File: hw/cevallos/u3/hw29Singleton/JavaScript/src/models/DiscountSingleton.js
const fs = require('fs');
const path = require('path');

class DiscountSingleton {
  constructor() {
    if (DiscountSingleton.instance) {
      return DiscountSingleton.instance;
    }
    
    this._configFile = path.join(__dirname, '../config/settings.json');
    this._discountPercentage = 0.0;
    this._initializeDiscount();
    
    DiscountSingleton.instance = this;
  }

  _initializeDiscount() {
    try {
      if (fs.existsSync(this._configFile)) {
        const configData = fs.readFileSync(this._configFile, 'utf8');
        const config = JSON.parse(configData);
        this._discountPercentage = parseFloat(config.discount_percentage) || 0.0;
      } else {
        // Create default config file
        this._saveConfig();
      }
    } catch (error) {
      console.error(`Error loading configuration: ${error.message}. Using default discount 0%.`);
      this._discountPercentage = 0.0;
    }
  }

  _saveConfig() {
    try {
      const configDir = path.dirname(this._configFile);
      if (!fs.existsSync(configDir)) {
        fs.mkdirSync(configDir, { recursive: true });
      }
      
      const config = {
        discount_percentage: this._discountPercentage
      };
      
      fs.writeFileSync(this._configFile, JSON.stringify(config, null, 2), 'utf8');
      return true;
    } catch (error) {
      console.error(`Error saving configuration: ${error.message}`);
      return false;
    }
  }

  getDiscountPercentage() {
    return this._discountPercentage;
  }

  setDiscountPercentage(newDiscount) {
    if (typeof newDiscount !== 'number' || newDiscount < 0 || isNaN(newDiscount)) {
      throw new Error('Discount must be a non-negative number.');
    }
    
    this._discountPercentage = newDiscount;
    return this._saveConfig();
  }

  calculateDiscountedPrice(originalPrice) {
    if (typeof originalPrice !== 'number' || originalPrice < 0) {
      throw new Error('Price must be a non-negative number.');
    }
    
    const discountAmount = originalPrice * (this._discountPercentage / 100);
    return Math.max(0, originalPrice - discountAmount);
  }

  static getInstance() {
    if (!DiscountSingleton.instance) {
      DiscountSingleton.instance = new DiscountSingleton();
    }
    return DiscountSingleton.instance;
  }
}

module.exports = DiscountSingleton;

File: hw/cevallos/u3/hw29Singleton/python/models/discount_singleton.py
import json
from typing import Any, Dict
from threading import Lock


class DiscountSingleton:
    """Singleton class to manage global discount percentage with thread safety."""
    
    _instance = None
    _lock: Lock = Lock()
    _config_file: str = "python/config/settings.json"
    _discount_percentage: float = 0.0
    
    def __new__(cls) -> "DiscountSingleton":
        """Create or return the singleton instance with thread safety."""
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._initialize_discount()
        return cls._instance
    
    def _initialize_discount(self) -> None:
        """Initialize discount from JSON configuration file."""
        try:
            with open(self._config_file, 'r') as file:
                config: Dict[str, Any] = json.load(file)
                self._discount_percentage = float(config.get("discount_percentage", 0.0))
        except (FileNotFoundError, json.JSONDecodeError, KeyError) as e:
            print(f"Error loading configuration: {e}. Using default discount 0%.")
            self._discount_percentage = 0.0
    
    def get_discount_percentage(self) -> float:
        """Get the current discount percentage."""
        return self._discount_percentage
    
    def set_discount_percentage(self, new_discount: float) -> bool:
        """Set new discount percentage and update JSON file."""
        if not isinstance(new_discount, (int, float)) or new_discount < 0:
            raise ValueError("Discount must be a non-negative number.")
        
        self._discount_percentage = float(new_discount)
        
        # Persist to JSON file
        try:
            with open(self._config_file, 'w') as file:
                json.dump({"discount_percentage": self._discount_percentage}, file, indent=4)
            return True
        except Exception as e:
            print(f"Error saving configuration: {e}")
            return False
    
    def calculate_discounted_price(self, original_price: float) -> float:
        """Calculate price after applying discount."""
        if original_price < 0:
            raise ValueError("Price cannot be negative.")
        
        discount_amount = original_price * (self._discount_percentage / 100)
        return max(0.0, original_price - discount_amount)